iroh-bytes/src/store/bao_file.rs

```diff
@@ -0,0 +1,1238 @@
+                io::ErrorKind::InvalidData,
+                "offset not aligned",
+            ));
+        }
+        let size_offset = (self.offset >> IROH_BLOCK_SIZE.0) << 3;
+        target.write_all_at(size_offset, self.size.to_le_bytes().as_slice())?;
+        Ok(())
+    }
+
+    /// The current size, representing the most correct size we know.
+    pub fn current_size(&self) -> u64 {
+        self.size
+    }
+
+    /// Convert to a vec in slot format.
+    pub fn to_vec(&self) -> io::Result<Vec<u8>> {
+        let mut res = Vec::new();
+        self.persist(&mut res)?;
+        Ok(res)
+    }
+}
+
+impl MutableMemStorage {
+    /// Get the parts data, outboard and sizes
+    pub fn into_parts(self) -> (SparseMemFile, SparseMemFile, SizeInfo) {
+        (self.data, self.outboard, self.sizes)
+    }
+
+    /// Create a new mutable mem storage from the given data
+    pub fn complete(bytes: Bytes) -> (Self, iroh_base::hash::Hash) {
+        let (outboard, hash) = raw_outboard(bytes.as_ref());
+        let res = Self {
+            data: bytes.to_vec().into(),
+            outboard: outboard.into(),
+            sizes: SizeInfo::complete(bytes.len() as u64),
+        };
+        (res, hash)
+    }
+
+    /// Persist the batch to disk, creating a FileBatch.
+    fn persist(&self, paths: DataPaths) -> io::Result<FileStorage> {
+        let mut data = create_read_write(&paths.data)?;
+        let mut outboard = create_read_write(&paths.outboard)?;
+        let mut sizes = create_read_write(&paths.sizes)?;
+        self.data.persist(&mut data)?;
+        self.outboard.persist(&mut outboard)?;
+        self.sizes.persist(&mut sizes)?;
+        data.sync_all()?;
+        outboard.sync_all()?;
+        sizes.sync_all()?;
+        Ok(FileStorage {
+            data,
+            outboard,
+            sizes,
+        })
+    }
+
+    pub(super) fn current_size(&self) -> u64 {
+        self.sizes.current_size()
+    }
+
+    pub(super) fn read_data_at(&self, offset: u64, len: usize) -> Bytes {
+        copy_limited_slice(&self.data, offset, len)
+    }
+
+    pub(super) fn data_len(&self) -> u64 {
+        self.data.len() as u64
+    }
+
+    pub(super) fn read_outboard_at(&self, offset: u64, len: usize) -> Bytes {
+        copy_limited_slice(&self.outboard, offset, len)
+    }
+
+    pub(super) fn outboard_len(&self) -> u64 {
+        self.outboard.len() as u64
+    }
+
+    pub(super) fn write_batch(
+        &mut self,
+        size: u64,
+        batch: &[BaoContentItem],
+    ) -> std::io::Result<()> {
+        let tree = BaoTree::new(ByteNum(size), IROH_BLOCK_SIZE);
+        for item in batch {
+            match item {
+                BaoContentItem::Parent(parent) => {
+                    if let Some(offset) = tree.pre_order_offset(parent.node) {
+                        let o0 = offset
+                            .checked_mul(64)
+                            .expect("u64 overflow multiplying to hash pair offset");
```

Comment:
why panic, instead of returning an error?

Reply:
this error would never happen. Even for a tree with data size u64::max, the offset will be something like data_size / chunk_group_size, so way below u64::max.

Maybe just using checked_mul here is overly paranoid and it should just be * 64.